victor/src/layout/mod.rs

```diff
@@ -343,6 +671,68 @@ impl TextRun {
     }
 }
 
+struct FlatVec<T>(Vec<T>);
+
+impl<T> Default for FlatVec<T> {
+    fn default() -> Self {
+        Self(vec![])
+    }
+}
+
+impl<T> From<Vec<T>> for FlatVec<T> {
+    fn from(vec: Vec<T>) -> Self {
+        Self(vec)
+    }
+}
+
+impl<'a, T> IntoIterator for FlatVec<T>
+where
+    T: Send + 'a,
+{
+    type Item = T;
+    type IntoIter = <Vec<T> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'a, T> IntoIterator for &'a mut FlatVec<T>
+where
+    T: Send + 'a,
+{
+    type Item = &'a mut T;
+    type IntoIter = <&'a mut Vec<T> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        (&mut self.0).into_iter()
+    }
+}
+
+impl<T> IntoParallelIterator for FlatVec<T>
+where
+    T: Send,
+{
+    type Iter = <Vec<T> as IntoParallelIterator>::Iter;
+    type Item = T;
+
+    fn into_par_iter(self) -> Self::Iter {
+        self.0.into_par_iter()
+    }
+}
+
+impl<T> ParallelExtend<Self> for FlatVec<T>
+where
+    T: Send,
+{
+    fn par_extend<I>(&mut self, par_iter: I)
+    where
+        I: IntoParallelIterator<Item = Self>,
+    {
+        self.0.par_extend(par_iter.into_par_iter().flatten());
```

Comment:
This is going to be a bit painful with `flatten()`, as rayon will split up those inner iterators (over `Self` = `FlatVec<T>`) into parallel chunks, and then the `Vec::par_extend` will collect those back into new temporary allocations, before it can finally fill the destination buffer in serial. Someday, I'd love to have rayon specialize this, similar to [what `Vec::from_iter` does](https://github.com/rust-lang/rust/blob/acd8dd6a50d505057a7d7ad8d0d7a4c2bd274200/src/liballoc/vec.rs#L1866-L1883), but for now the generic code can't deal with this very well.

Instead, you can shortcut [what rayon is doing](https://github.com/rayon-rs/rayon/blob/003b5e64735a6606c0aa7735cc22f0a259056c2f/src/iter/collect/mod.rs#L140-L159) yourself, something like:

```rust
let list: LinkedList<FlatVec<T>> = par_iter.into_par_iter().collect();
self.0.reserve(list.iter().map(Vec::len).sum());
for mut flat_vec in list {
    self.0.append(&mut flat_vec.0);
}
```

You could also use the `opt_len()` trick for indexed iterators to collect `Vec<Self>` instead of `LinkedList` before flattening, but in the grand scheme this is probably a small optimization.

Reply:
Just to be sure about what you just said: at no point Rayon will allocate a `Vec<FlatVec<T>>` itself, right?

Reply:
If you leave it to rayon `flatten().collect::<Vec<T>>()`, it will collect a temporary `LinkedList<Vec<T>>`, folding into new vectors and reducing in a list. Then it will serially write that into the final `Vec<T>`.